Reorganization of best practices and addition of overlays for globals (#201)
OAI/Documentation: best-practices/describing-large-apis.md

---
layout: default
title: Describing Large APIs
parent: Best Practices
nav_order: 6
---

> # Describing Large APIs

Reviewer: ```suggestion
# Describe Large APIs
```

This is a collection of small hints related to working with large OADs.

> - **Do not repeat yourself** (The DRY principle). If the same piece of YAML or JSON appears more than once in the document, it's time to move it to the `components` section and reference it from other places using `$ref` (See [Reusing Descriptions](../specification/components). Not only will the resulting document be smaller but it will also be much easier to maintain).

Reviewer 2: Change "more than once" to "multiple times" - just repeating the same content more than once is not usually enough to merit factoring it out into a a reusable component, if readability is diminished.  When to factor out common code is always subjective.

Author: ```suggestion
- **Do not repeat yourself** (The DRY principle). If the same piece of YAML or JSON appears multiple times in the document, it's time to move it to the `components` section and reference it from other places using `$ref` (See [Reusing Descriptions](../specification/components). Not only will the resulting document be smaller but it will also be much easier to maintain).
```

  Components can be referenced from other documents, so you can even reuse them across different API descriptions!

- **Split the description into several documents**: Smaller files are easier to navigate, but too many of them are equally taxing. The key lies somewhere in the middle.

>   A good rule of thumb is to use the natural hierarchy present in URLs to build your directory structure. For example, put all routes starting with `/users` (like `/users` and `/users/{id}`) in the same file (think of it as a "sub-API").

Reviewer 2: Switch off of colloquial and potentially offensive expression:

```suggestion
  A good principle is to use the natural hierarchy present in URLs to build your directory structure. For example, put all routes starting with `/users` (like `/users` and `/users/{id}`) in the same file (think of it as a "sub-API").
```
